### PR TITLE
perf(monitoring-daemon): Increase the performance of the _snakeify function

### PR DESCRIPTION
**Pull request by dwilliams-armory**
perf(monitoring-daemon): Increase the performance of the _snakeify function

This replaces the `_snakeify` function with a more performant version, which removes the regex (not very performant in python) and adds a small cache dictionary.  This won't fix the socket issue in https://github.com/spinnaker/spinnaker/issues/5373, but should help prevent the timeouts that trigger the socket issue.

### CODE REVIEW

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py [hunk 1]
```diff
@@ -25,15 +25,24 @@
...
 import yaml
 
-
-# see https://stackoverflow.com/questions/1175208/\
-#     elegant-python-function-to-convert-camelcase-to-snake-case
-_SNAKE_CASE_RE = re.compile('((?<=[a-z0-9])[A-Z]|(?!^)(?<!_)[A-Z](?=[a-z]))')
-
+# This is a cache for the _snakeify function to increase performance on the
+#    conversion of CamelCase to snake_case
+SNAKEIFY_CACHE = {}
```
Comment (ezimanyi) [inline spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:30]: I'm curious how much this cache helps vs. the risk of having a memory leak here.  Given that there's no expiration here, this will grow without bound; while it's only holding strings I do still worry about having this here.  If we need to memoize, is there maybe a library we can use that limits the size of the cache?

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py [hunk 2]
```diff
@@ -25,15 +25,24 @@
...
 
 def _snakeify(text):
-  """Turn text into snake-case."""
-  return _SNAKE_CASE_RE.sub(r'_\1', text).lower()
+  if text not in SNAKEIFY_CACHE.keys():
```
Comment (ezimanyi) [inline spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:33]: Unless python is doing some short-circuiting here, this lookup will be `O(N)` in the number of keys in the cache, which will get worse and worse as the monitoring daemon has been running longer, potentially eliminating any benefit of caching.

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py (hunk 1 above)
Comment (dwilliams-armory) [reply spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:30]: When I ran the unittest 1,000,000 times, the regex took about 44 seconds, the function without cache took about 28 seconds, and the cache took about 2.5 seconds.  I'm also not as concerned with the cache growing too much as the names and labels that this function is being used for is from the spectator monitoring endpoint, which should have a finite count and be repeated every 30 or so seconds when pulling the mentrics to (in our clients case) prometheus.

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py (hunk 2 above)
Comment (dwilliams-armory) [reply spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:33]: As the names and labels should be defined in the service code, and repeated on each lookup, this shouldn't be an issue.  I'd be happy to add some sort of aging out mechanism if it proves to be a problem.

**Comment by dwilliams-armory** [comment]
I'll have a look at that library.  The reasons I haven't used a library are that the libraries I've seen include lots of other code that does appear to be necessary for a simple task, and that all the libraries that I checked used regex under the hood.  I've been specifically trying to avoid regexs in python as they're terribly non-performant.

**Comment by link108** [comment]
@dwilliams-armory do you know the domain of input values for that function? I see the tests pass, but strings ending with a a lowercase letter and then a capital letter don’t get snakified correctly.
Eg: `hiHiB` should be `hi_hi_b`, but this function returns `hi_hib` . If you add another character to the end, it works: `hiHiBi` is correctly transformed to `hi_hi_bi`. `hiHiBB` is correctly transformed to `hi_hi_bb`, so this only applies to strings with the ending character being uppercase

from what I can see, the issue is caused by the for loop range:
`for position in range(1, len(text) - 1)`
which doesn’t look at the last letter, so that
`text[position + 1].islower()`
can succeed and not give out of bounds error.

I think the range can be fixed to be range(1, len(text)) and there should be an additional check to ensure text[position + 1] is within the string bounds.

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py (hunk 1 above)
Comment (ezimanyi) [reply spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:30]: I think this is a case where the performance of repeating the unit test might not be a great proxy for what would happen in production...repeating the test with the same keys would mean a 99.9999% hit rate of the cache.

Obviously if that's similar to the cache hit rate in production then this would be a good proxy...do you have an idea for the typical number of keys that will be here, and how often they change?  If this is a static list of keys that don't change, this is probably reasonable, but just want to make sure we check that assumption independently.

Comment (ezimanyi) [reply spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:30]: I think this is a case where the performance of repeating the unit test might not be a great proxy for what would happen in production...repeating the test with the same keys would mean a 99.9999% hit rate of the cache.

Obviously if that's similar to the cache hit rate in production then this would be a good proxy...do you have an idea for the typical number of keys that will be here, and how often they change?  If this is a static list of keys that don't change, this is probably reasonable, but just want to make sure we check that assumption independently.

Comment (dwilliams-armory) [reply spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:30]: This is the only area in the code that the `_snakeify` function is used: https://github.com/spinnaker/spinnaker-monitoring/blob/a724ee721df36d347530313e36c0dab241126c67/spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py#L929
It appears to only be used to convert metric labels and metric names, which should be coming from the code in the Spinnaker service being monitored (e.g. clouddriver).  If metric labels and names are constantly changing, then monitoring would not be effective.  Please verify this, but as far as I can tell, this will be a set list of labels and name that will be created initially, on the first request or two, then come from the cache table afterwards for the rest of the time the daemon is running.

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py (hunk 2 above)
Comment (dwilliams-armory) [reply spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py]: Looks like the option @ezimanyi suggested is the most performant, so used that one.

**Comment by dwilliams-armory** [comment]
Here's a script that I used, based off the unittest, to profile this PR:

**Comment by german-muzquiz** [comment]
My two cents: How about using the existing regex + caching? If the metric names are the same over and over again, only the first pass would take long. It's just for the sake of using a library/known method instead of writing our own code for this conversion.

**Comment by ezimanyi** [comment]
Thanks for the detailed analysis!  I think I'm convinced at this point that it makes sense to cache this, so the only remaining question is whether to cache the existing implementation or a new one.  I'm slightly leaning towards @german-muzquiz 's suggestion to just cache the existing implementation as it's battle-tested, but I also am not a huge fan of regular expressions so won't complain to see one go away 🤷‍♂

**Comment by dwilliams-armory** [comment]
I'm not a fan of regex either, so I'd prefer to not use the regex.  I am fine with converting it to caching with regex if the consensus here leans that way.

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py [hunk 3]
```diff
@@ -25,15 +25,23 @@
...
-  """Turn text into snake-case."""
-  return _SNAKE_CASE_RE.sub(r'_\1', text).lower()
+  if text not in SNAKEIFY_CACHE:
+    result = []
+    result.append(text[0].lower())
+    for position in range(1, len(text)):
+      if text[position].isupper():
+        if text[position - 1].islower():
+          result.append('_')
+        elif position < (len(text) - 1) and text[position + 1].islower() and text[position - 1] is not '_':
```
Comment (ezimanyi) [inline spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:40]: Maybe a micro-optimization...but we're doing `len(text)` on every iteration; could this be a local variable?

**Comment by ezimanyi** [comment]
I'd probably vote in order: (1) use a library, (2) use the custom code here, (3) use the regex...only because there are bound to be a lot of edge cases (one of which was already found) and the unit tests aren't super robust so I'd probably rather use a library (even if it pulls in some other stuff).

**Comment by dwilliams-armory** [comment]
While I would agree with your plan and reasoning, I haven't found a library yet that doesn't use a regex.  Even the library mentioned above uses regex under the hood.

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py (hunk 3 above)
Comment (dwilliams-armory) [reply spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py]: Added this micro-optimization.

File: spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py [hunk 4]
```diff
@@ -25,15 +25,24 @@
...
+# This is a cache for the _snakeify function to increase performance on the
+#    conversion of CamelCase to snake_case
+SNAKEIFY_CACHE = {}
 
 def _snakeify(text):
-  """Turn text into snake-case."""
-  return _SNAKE_CASE_RE.sub(r'_\1', text).lower()
+  if text not in SNAKEIFY_CACHE:
+    result = []
+    result.append(text[0].lower())
```
Comment (link108) [inline spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py:35]:
```suggestion
    result = [text[0].lower()]
```
